File: src/algorithms/pathfinding/flood-fill/flood-fill-bfs/__tests__/flood-fill-bfs.test.ts
import { describe, it, expect } from "vitest";
import type { GridCell } from "@/types";
import { floodFillBfs } from "../sources/flood-fill-bfs.ts?fn";

function createEmptyGrid(rows: number, cols: number): GridCell[][] {
  return Array.from({ length: rows }, (_, rowIndex) =>
    Array.from({ length: cols }, (_, colIndex) => ({
      row: rowIndex,
      col: colIndex,
      type: "empty" as const,
      state: "default" as const,
    })),
  );
}

function setCell(grid: GridCell[][], row: number, col: number, type: GridCell["type"]): void {
  const gridRow = grid[row];
  if (gridRow) {
    const cell = gridRow[col];
    if (cell) cell.type = type;
  }
}

describe("floodFillBfs", () => {
  it("fills all cells on a small empty grid", () => {
    const grid = createEmptyGrid(3, 3);
    setCell(grid, 0, 0, "start");

    const result = floodFillBfs(grid, [0, 0]);

    expect(result.count).toBe(9);
    expect(result.filled.length).toBe(9);
  });

  it("respects walls and does not fill through them", () => {
    const grid = createEmptyGrid(3, 3);
    /* Wall divides left column from the rest */
    setCell(grid, 0, 1, "wall");
    setCell(grid, 1, 1, "wall");
    setCell(grid, 2, 1, "wall");

    const result = floodFillBfs(grid, [0, 0]);

    /* Only left column (3 cells) should be filled */
    expect(result.count).toBe(3);
  });

  it("fills only the connected region when surrounded by walls", () => {
    const grid = createEmptyGrid(5, 5);
    /* Enclose a 3x3 region in the center */
    for (let col = 0; col < 5; col++) {
      setCell(grid, 0, col, "wall");
      setCell(grid, 4, col, "wall");
    }
    for (let row = 1; row < 4; row++) {
      setCell(grid, row, 0, "wall");
      setCell(grid, row, 4, "wall");
    }

    const result = floodFillBfs(grid, [2, 2]);

    /* Interior 3x3 = 9 cells */
    expect(result.count).toBe(9);
  });

  it("starts with the seed cell in the filled list", () => {
    const grid = createEmptyGrid(3, 3);

    const result = floodFillBfs(grid, [1, 1]);

    expect(result.filled[0]).toEqual([1, 1]);
  });

  it("fills a single isolated cell when all neighbors are walls", () => {
    const grid = createEmptyGrid(3, 3);
    setCell(grid, 0, 1, "wall");
    setCell(grid, 1, 0, "wall");
    setCell(grid, 1, 2, "wall");
    setCell(grid, 2, 1, "wall");

    const result = floodFillBfs(grid, [1, 1]);

    expect(result.count).toBe(1);
    expect(result.filled[0]).toEqual([1, 1]);
  });

  it("count matches filled array length", () => {
    const grid = createEmptyGrid(4, 4);
    setCell(grid, 2, 0, "wall");
    setCell(grid, 2, 1, "wall");
    setCell(grid, 2, 2, "wall");

    const result = floodFillBfs(grid, [0, 0]);

    expect(result.count).toBe(result.filled.length);
  });

  it("treats start and end cell types as passable", () => {
    const grid = createEmptyGrid(3, 3);
    setCell(grid, 0, 0, "start");
    setCell(grid, 2, 2, "end");

    const result = floodFillBfs(grid, [0, 0]);

    /* All 9 cells should be filled including start and end */
    expect(result.count).toBe(9);
  });
});

File: src/algorithms/pathfinding/flood-fill/flood-fill-bfs/__tests__/FloodFillBfs_test.cpp
#include "../sources/FloodFillBfs.cpp"
#include <cassert>
#include <iostream>

std::vector<std::vector<GridCell>> makeEmptyGrid(int rows, int cols) {
  std::vector<std::vector<GridCell>> grid(rows, std::vector<GridCell>(cols));
  for (int row = 0; row < rows; row++) {
    for (int col = 0; col < cols; col++) {
      grid[row][col] = {row, col, CellType::Empty, "default"};
    }
  }
  return grid;
}

void setWall(std::vector<std::vector<GridCell>>& grid, int row, int col) {
  grid[row][col].cellType = CellType::Wall;
}

int main() {
  // Test: fills all cells on small empty grid
  {
    auto grid = makeEmptyGrid(3, 3);
    auto result = floodFillBfs(grid, {0, 0});
    assert(result.count == 9);
    assert((int)result.filled.size() == 9);
  }

  // Test: respects walls
  {
    auto grid = makeEmptyGrid(3, 3);
    setWall(grid, 0, 1);
    setWall(grid, 1, 1);
    setWall(grid, 2, 1);
    auto result = floodFillBfs(grid, {0, 0});
    assert(result.count == 3);
  }

  // Test: enclosed region
  {
    auto grid = makeEmptyGrid(5, 5);
    for (int col = 0; col < 5; col++) {
      setWall(grid, 0, col);
      setWall(grid, 4, col);
    }
    for (int row = 1; row < 4; row++) {
      setWall(grid, row, 0);
      setWall(grid, row, 4);
    }
    auto result = floodFillBfs(grid, {2, 2});
    assert(result.count == 9);
  }

  // Test: seed cell is first filled
  {
    auto grid = makeEmptyGrid(3, 3);
    auto result = floodFillBfs(grid, {1, 1});
    assert(result.filled[0].first == 1 && result.filled[0].second == 1);
  }

  // Test: isolated cell
  {
    auto grid = makeEmptyGrid(3, 3);
    setWall(grid, 0, 1);
    setWall(grid, 1, 0);
    setWall(grid, 1, 2);
    setWall(grid, 2, 1);
    auto result = floodFillBfs(grid, {1, 1});
    assert(result.count == 1);
    assert(result.filled[0].first == 1 && result.filled[0].second == 1);
  }

  // Test: count matches filled length
  {
    auto grid = makeEmptyGrid(4, 4);
    setWall(grid, 2, 0);
    setWall(grid, 2, 1);
    setWall(grid, 2, 2);
    auto result = floodFillBfs(grid, {0, 0});
    assert(result.count == (int)result.filled.size());
  }

  std::cout << "All tests passed!" << std::endl;
  return 0;
}
